iris-mpc-cpu/src/shares/share.rs

```diff
@@ -319,3 +319,20 @@ impl<T: IntRing2k> Shl<u32> for Share<T> {
         }
     }
 }
+
+// Additive share of a Hamming distance value
+#[derive(Debug, Clone, Hash, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(bound = "")]
+pub struct DistanceShare<T: IntRing2k> {
+    pub code_dot: Share<T>,
+    pub mask_dot: Share<T>,
```

Comment:
Yes, I guess this involves two separate design questions: do we store distance values that are bit-lifted, and do we store replicated or non-replicated shares.

To me preprocessing the bit lifting step seems like a no-brainer: the lifting procedure has to happen at least once for each distance that is computed (I don't think any distances are stored in the graph database which are not also used in a distance comparison operation), so paying a memory storage premium seems worthwhile to avoid repeating the fairly heavy multi-round lifting procedure for distances that have already been computed once.  Concrete metrics for how often a particular distance is used in these comparisons are one of my near-term tasks, but I would guess that on average cached distances will be compared at least a couple of times during their tenancy in the HNSW graph.

The storage of the replicated vs. non-replicated shares is not as clear to me -- storing replicated shares increases the memory requirements by the same factor as the bit-lifting preprocessing in exchange for one round of communication to reshare when a distance comparison is done.  Since distance comparison already requires a 32-bit high-order-bit extraction protocol, the single extra communication round doesn't seem like such a big deal in this context to me (especially compared to the bit-lifting which involves a nontrivial number of communication rounds).

At any rate, definitely seems like a trade-off space we should think about -- the big difference here as compared with MPCv2 being that distances will generally be involved in multiple comparison operations after being computed.

Reply:
@dkales in response to your question about the number of distances that are cached, it's quite a lot -- each edge of the HNSW graph has an associated cached distance, so we would expect 100 to 200 of these to be stored for each iris code in the system.  So the space requirements for the HNSW graph with these cached distances is nontrivial, but perhaps sitting at 5-15% of the space needed for the secret shared iris codes themselves.